src/main/java/org/truffleruby/core/regexp/RegexpNodes.java

```diff
@@ -118,41 +123,60 @@ private static Regex makeRegexpForEncoding(RubyContext context, DynamicObject re
         return regex;
     }
 
-    public static DynamicObject matchCommon(RubyContext context, Node currentNode, DynamicObject regexp, DynamicObject string, int startPos, boolean onlyMatchAtStart) {
-        final Matcher matcher = createMatcher(context, regexp, string, true);
-        int range = StringOperations.rope(string).byteLength();
-        return matchCommon(context, currentNode, regexp, string, matcher, startPos, range, onlyMatchAtStart);
-    }
+    @NodeChildren({
+            @NodeChild(type = RubyNode.class, value = "regexp"),
+            @NodeChild(type = RubyNode.class, value = "source"),
+            @NodeChild(type = RubyNode.class, value = "matcher"),
+            @NodeChild(type = RubyNode.class, value = "startPos"),
+            @NodeChild(type = RubyNode.class, value = "range"),
+            @NodeChild(type = RubyNode.class, value = "onlyAtStart")
+    })
+    public static abstract class MatchNode extends RubyNode {
+        @Child private TaintResultNode taintResultNode = new TaintResultNode();
+        @Child private AllocateObjectNode allocateNode = AllocateObjectNode.create();
 
-    public static DynamicObject matchCommon(RubyContext context, Node currentNode, DynamicObject regexp, DynamicObject string, Matcher matcher,
-            int startPos, int range, boolean onlyMatchAtStart) {
-        assert RubyGuards.isRubyRegexp(regexp);
-        assert RubyGuards.isRubyString(string);
+        public static MatchNode create() {
+            return MatchNodeGen.create(null, null, null, null, null, null);
+        }
 
-        int match = runMatch(context, currentNode, matcher, startPos, range, onlyMatchAtStart);
+        public abstract DynamicObject execute(DynamicObject regexp, DynamicObject string, Matcher matcher,
+                int startPos, int range, boolean onlyMatchAtStart);
 
-        if (match == -1) {
-            return context.getCoreLibrary().getNil();
-        }
+        @Specialization
+        public DynamicObject execute(DynamicObject regexp, DynamicObject string, Matcher matcher,
```

Comment:
specializations should be `protected`, not `public`, and should not start with `execute` to avoid confusion.